flang/lib/Optimizer/Passes/Pipelines.cpp

```diff
@@ -227,11 +227,11 @@ void createHLFIRToFIRPassPipeline(mlir::PassManager &pm, bool enableOpenMP,
                                          hlfir::createOptimizedBufferization);
   }
   pm.addPass(hlfir::createLowerHLFIROrderedAssignments());
+  if (enableOpenMP)
+    pm.addPass(flangomp::createLowerWorkshare());
```

Comment:
I think the workshare lowering pass assumes it runs after HLFIR bufferization. For example see the changes in the previous PR https://github.com/llvm/llvm-project/pull/104748. @ivanradanov please could you advise on this.

Another approach would be to modify `createLowerHLFIRIntrinsics` so that it does not lower eligable `hlfir.sum` operations if they are inside of a workshare (a bit like how bufferization was modified in the PR above).

Reply:
I think we can use the code that lowers `hlfir.<intrinsic>` into openmp loops at the point the `hlfir.<intrinsic>`s are lowered into `fir.call`s into the flang rt library. We should have a check for whether we should use the generic lowering or the openmp loop lowering (using the `shouldUseWorkshareLowering` function) and then for the supported intrinsics lower it to openmp loops instead of the rt call.

Another option is to have another pass which runs before the pass that does the `fir.call` lowering and that pass takes care of the intrinsics we support.

Even if we use the first approach we should still probably have a separate compilation unit for the collection of those. This way it should not interfere much with the existing code and should be easy to integrate.